// mocha/lib/reporters/json.js
/**
 * @module JSON
 */
/**
 * Module dependencies.
 */

const fs = require('fs');
const path = require('path');
const Base = require('./base');
const { createUnsupportedError } = require('../errors');
const utils = require('../utils');
const { constants } = require('../runner');

const { EVENT_TEST_PASS } = constants;
const { EVENT_TEST_PENDING } = constants;
const { EVENT_TEST_FAIL } = constants;
const { EVENT_TEST_END } = constants;
const { EVENT_RUN_END } = constants;

/**
 * Expose `JSON`.
 */

exports = module.exports = JSONReporter;

/**
 * Constructs a new `JSON` reporter instance.
 *
 * @public
 * @class JSON
 * @memberof Mocha.reporters
 * @extends Mocha.reporters.Base
 * @param {Runner} runner - Instance triggers reporter actions.
 * @param {Object} [options] - runner options
 */
function JSONReporter(runner, options = {}) {
  Base.call(this, runner, options);

  const self = this;
  const tests = [];
  const pending = [];
  const failures = [];
  const passes = [];
  let output;

  if (options.reporterOption && options.reporterOption.output) {
    if (utils.isBrowser()) {
      throw createUnsupportedError('file output not supported in browser');
    }
    output = options.reporterOption.output;
  }

  runner.on(EVENT_TEST_END, (test) => {
    tests.push(test);
  });

  runner.on(EVENT_TEST_PASS, (test) => {
    passes.push(test);
  });

  runner.on(EVENT_TEST_FAIL, (test) => {
    failures.push(test);
  });

  runner.on(EVENT_TEST_PENDING, (test) => {
    pending.push(test);
  });

  runner.once(EVENT_RUN_END, () => {
    const obj = {
      stats: self.stats,
      tests: tests.map(clean),
      pending: pending.map(clean),
      failures: failures.map(clean),
      passes: passes.map(clean),
    };

    runner.testResults = obj;

    const json = JSON.stringify(obj, null, 2);
    if (output) {
      try {
        fs.mkdirSync(path.dirname(output), { recursive: true });
        fs.writeFileSync(output, json);
      } catch (err) {
        console.error(
          `${Base.symbols.err} [mocha] writing output to "${output}" failed: ${err.message}\n`,
        );
        process.stdout.write(json);
      }
    } else {
      process.stdout.write(json);
    }
  });
}

/**
 * Return a plain-object representation of `test`
 * free of cyclic properties etc.
 *
 * @private
 * @param {Object} test
 * @return {Object}
 */
function clean(test) {
  let err = test.err || {};
  if (err instanceof Error) {
    err = errorJSON(err);
  }

  return {
    title: test.title,
    fullTitle: test.fullTitle(),
    file: test.file,
    duration: test.duration,
    currentRetry: test.currentRetry(),
    speed: test.speed,
    err: cleanCycles(err),
  };
}

/**
 * Replaces any circular references inside `obj` with '[object Object]'
 *
 * @private
 * @param {Object} obj
 * @return {Object}
 */
function cleanCycles(obj) {
  const cache = [];
  return JSON.parse(
    JSON.stringify(obj, (key, value) => {
      if (typeof value === 'object' && value !== null) {
        if (cache.indexOf(value) !== -1) {
          // Instead of going in a circle, we'll print [object Object]
          return `${value}`;
        }
        cache.push(value);
      }

      return value;
    }),
  );
}

/**
 * Transform an Error object into a JSON object.
 *
 * @private
 * @param {Error} err
 * @return {Object}
 */
function errorJSON(err) {
  const res = {};
  Object.getOwnPropertyNames(err).forEach((key) => {
    res[key] = err[key];
  }, err);
  return res;
}

JSONReporter.description = 'single JSON object';
