// mocha/lib/reporters/json-stream.js
/**
 * @module JSONStream
 */
/**
 * Module dependencies.
 */

const Base = require('./base');
const { constants } = require('../runner');

const { EVENT_TEST_PASS } = constants;
const { EVENT_TEST_FAIL } = constants;
const { EVENT_RUN_BEGIN } = constants;
const { EVENT_RUN_END } = constants;

/**
 * Expose `JSONStream`.
 */

exports = module.exports = JSONStream;

/**
 * Constructs a new `JSONStream` reporter instance.
 *
 * @public
 * @class
 * @memberof Mocha.reporters
 * @extends Mocha.reporters.Base
 * @param {Runner} runner - Instance triggers reporter actions.
 * @param {Object} [options] - runner options
 */
function JSONStream(runner, options) {
  Base.call(this, runner, options);

  const self = this;
  const { total } = runner;

  runner.once(EVENT_RUN_BEGIN, () => {
    writeEvent(['start', { total }]);
  });

  runner.on(EVENT_TEST_PASS, (test) => {
    writeEvent(['pass', clean(test)]);
  });

  runner.on(EVENT_TEST_FAIL, (test, err) => {
    test = clean(test);
    test.err = err.message;
    test.stack = err.stack || null;
    writeEvent(['fail', test]);
  });

  runner.once(EVENT_RUN_END, () => {
    writeEvent(['end', self.stats]);
  });
}

/**
 * Mocha event to be written to the output stream.
 * @typedef {Array} JSONStream~MochaEvent
 */

/**
 * Writes Mocha event to reporter output stream.
 *
 * @private
 * @param {JSONStream~MochaEvent} event - Mocha event to be output.
 */
function writeEvent(event) {
  process.stdout.write(`${JSON.stringify(event)}\n`);
}

/**
 * Returns an object literal representation of `test`
 * free of cyclic properties, etc.
 *
 * @private
 * @param {Test} test - Instance used as data source.
 * @return {Object} object containing pared-down test instance data
 */
function clean(test) {
  return {
    title: test.title,
    fullTitle: test.fullTitle(),
    file: test.file,
    duration: test.duration,
    currentRetry: test.currentRetry(),
    speed: test.speed,
  };
}

JSONStream.description = 'newline delimited JSON events';
